docs/features/source-generators.cookbook.md

```diff
@@ -0,0 +1,250 @@
+- Run _un-ordered_, each generator will see the same input compilation, with no access to files created by other source generators.
+- A user specifies the generators to run via list of assemblies, much like analyzers.
+
+## Out of scope designs
+
+We will briefly look at the non-solvable problems as examples of the kind of problems source generators are *not* designed to solve:
+
+### Language features
+
+Source generators are not designed to replace new language features: for instance one could imagine [records](records.md) being implemented as a source generator
+that converts the specified syntax to a compilable C# representation.
+
+We explicitly consider this to be an anti-pattern; the language will continue to evolve and add new features, and we don't expect source generators to be
+a way to enable this. Doing so would create new 'dialects' of C# that are incompatible with the compiler without generators. Further, because generators,
+by design, cannot interact with each other, language features implemented in this way would quickly become incompatible with other additions to the language.
+
+### Code-not-source generators
+
+There are many post-processing tasks that users perform on their assemblies today, which here we define as 'code-not-source' generation. These include, but
+are not limited to:
+
+- Optimization
+- Logging injection
+- IL Weaving
+- Call site re-writing
+
+While these techniques have many valuable use cases, they do not fit into the idea of *source* generation. They are, by defintion, code altering operations
+which is explicitly ruled out by the source generation proposal.
+
+There are already well supported tools and techniques for acheiving these kinds of operations, and the source generators proposal is not aimed at replacing them.
+
+## Designs
+
+This section is broken down by user scenarios, with general solutions listed first, and more specific examples later on.
+
+### Generated class
+
+**User scenario:** As a generator author I want to be able to create a class that can be referenced from user code.
+
+**Solution:** Generate the class based on the data available in the compilation, and add the new source to compilation.
+
+**Example:**
+
+```csharp
+public class CustomGenerator : ISourceGenerator
+{
+    public void Execute(SourceGeneratorContext context)
+    {
+        context.AddSource("myGeneratedFile.cs", SourceText.From($@"
+namespace GeneratedNamespace
+{
+    public class GeneratedClass
+    {
+        public static void GeneratedMethod()
+        {
+            // generated code
+        }
+    }
+}"));
+    }
+}
+```
+
+```csharp
+public partial class UserClass
+{
+    public void UserMethod()
+    {
+        // call into a generated method
+        GeneratedNamespace.GeneratedClass.GeneratedMethod();
+    }
+}
+```
+
+### Single file transformation
+
+**User scenario:** As a generator author I want to be able to transform a single file into a C# representation.
```

Comment:
I think we should be clearer that it's a non-C# file that is being transformed here. Casual reader may think you're translating a C# file.